core/src/main/java/org/lfenergy/shapeshifter/core/service/sending/UftpSendMessageService.java

```diff
@@ -33,114 +35,142 @@
+    private String getSignedXml(PayloadMessageType payloadMessage, SigningDetails details) {
+        var payloadXml = serializer.toXml(payloadMessage);
+        var signedMessage = cryptoService.signMessage(payloadXml, details.sender(), details.senderPrivateKey());
+        return serializer.toXml(signedMessage);
+    }
+
+    private void send(String signedXml, UftpParticipant recipient) {
+        var url = participantService.getEndPointUrl(recipient);
+        send(signedXml, url, MAX_FOLLOW_REDIRECTS);
+    }
+
+    private void send(String signedXml, String url, int maxFollowRedirects) {
+        try {
+            log.debug(String.format("Sending message to: %s", url));
+
+            var request = HttpRequest.newBuilder()
+                    .uri(new URI(url))
+                    .POST(BodyPublishers.ofString(signedXml))
+                    .setHeader("Content-Type", "text/xml")
+                    .build();
+
+            var response = httpClient.send(request, BodyHandlers.ofString());
+
+            var httpStatusCode = HttpStatusCode.getByValue(response.statusCode());
+
+            if (!httpStatusCode.isSuccess()) {
+                // According to the specification: redirects (responses with status code 3xx) should be honored in order to support load balancing
+                if (httpStatusCode.isRedirect() && FOLLOW_REDIRECT_STATUS_CODES.contains(httpStatusCode)) {
+                    if (maxFollowRedirects <= 0) {
+                        throw new UftpSendException(MessageFormat.format(MSG_TOO_MANY_REDIRECTS, url));
+                    }
+
+                    var redirectUrl = response.headers().firstValue(REDIRECT_LOCATION_HEADER_NAME)
+                            .orElseThrow(() -> new UftpServerErrorException(MessageFormat.format(MSG_MISSING_REDIRECT_LOCATION, url), httpStatusCode));
+
+                    send(signedXml, redirectUrl, maxFollowRedirects - 1);
+                } else if (httpStatusCode.isClientError()) {
+                    throw new UftpClientErrorException(MessageFormat.format(MSG_CLIENT_ERROR, response.statusCode(), url, response.body()), httpStatusCode);
+                } else if (httpStatusCode.isServerError()) {
+                    throw new UftpServerErrorException(MessageFormat.format(MSG_SERVER_ERROR, response.statusCode(), url, response.body()), httpStatusCode);
+                } else {
+                    throw new UftpSendException(MessageFormat.format(MSG_UNEXPECTED_RESPONSE_STATUS, response.statusCode(), url, response.body()), httpStatusCode);
+                }
+            }
+        } catch (URISyntaxException | IllegalArgumentException e) {
+            throw new UftpSendException(MessageFormat.format(MSG_INVALID_ENDPOINT, e.getMessage()), e);
+        } catch (IOException e) {
+            throw new UftpSendException(MessageFormat.format(MSG_UNEXPECTED_IO_ERROR, url, e.getMessage()), e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
one last question: why do we need `Thread.currentThread().interrupt()` - it suggests that this code is running on a separate thread or something?

Reply:
Indeed, the `httpClient.send(...)` call can throw an `InterruptedException`.

The addition of `Thread.currentThread().interrupt();` was to fix a SonarQube warning that we must always mark the current thread as interrupted when catching an `InterruptedException`

Reply:
Ah, ok...